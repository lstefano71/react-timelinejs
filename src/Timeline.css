/* axis */

/*.axis { !* axis labels *!*/
    /*fill: #808080;*/
    /*font-family: sans-serif;*/
    /*font-size: 10px;*/
/*}*/

/*.axis line{ !* axis tick marks *!*/
    /*stroke-width : 1;*/
    /*stroke: grey;*/
    /*shape-rendering: crispEdges;*/
/*}*/

/*.axis path { !* axis line *!*/
    /*stroke-width : 1;*/
    /*stroke: grey;*/
    /*shape-rendering: crispEdges;*/
/*}*/

/* timeline band */

.timeline {
    position: relative;
}

.band { /* band background */
    fill: #FAFAFA;
}

/* labels */

.bandLabel {
    fill: #F0F0F0;
    font: 10px sans-serif;
    font-weight: bold;
}

.bandMinMaxLabel {
    fill: blue;
    font: 10px sans-serif;
    font-weight: bold;
}

.bandMidLabel {
    fill: red;
    font: 10px sans-serif;
    font-style: italic;
    font-weight: bold;
}

/* brush */

.brush .extent {
    stroke: gray;
    fill: blue;
    fill-opacity: .1;
}

.chart {
    fill: #EEEEEE;
}

.interval {
    fill: dodgerblue;
    stroke-width: 6;
    cursor : default;
    pointer-events: true;
}

.instant {
    fill: dodgerblue;
    stroke-width: 6;
    cursor : default;
    /*pointer-events: true;*/
}

.instantLabel {
    fill : blue;
    font: 10px sans-serif;
    shape-rendering: crispEdges;
}

.intervalLabel {
    fill : black;
    font: 10px sans-serif;
    shape-rendering: crispEdges;
}

.item {
    cursor : default;
    pointer-events: auto;
}

/*.svg {*/
    /*border-style: solid;*/
    /*border-width: 1px;*/
    /*border-color: black;*/
    /*background-color: #FFFFFF;*/
/*}*/

.tooltip {
    width: 0px;
    height: 0px;
    position: absolute;
    z-index: 2;
    visibility: hidden;
    color : black;
    cursor:default;
    pointer-events: none;
    display: flex;
}

.tooltip.top-left {

}
.tooltip.top-right {
    flex-direction: row-reverse;
}
.tooltip.bottom-right {
    flex-flow: row-reverse wrap-reverse;
}
.tooltip.bottom-left {
    flex-flow: column-reverse;
}

.tooltip-container {
    shape-rendering: crispEdges;
    padding: 4px;
    min-width: 300px;
    min-height: 100px;
    border: 1px solid;
    background-color: #FFFFEE;
}

.mouseover-line {
    stroke: #0984e3;
}

.mouseover-label {
    fill : white;
    font: 10px sans-serif;
    shape-rendering: crispEdges;
    text-anchor: middle;
}

.mouseover-label-background {
    fill: #0984e3;
}
