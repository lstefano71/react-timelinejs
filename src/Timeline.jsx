import React from 'react';
import timeline from './TimelineModel';
import './Timeline.css';
import equals from 'ramda/src/equals';
import PropTypes from 'prop-types';

const propTypes = {
    data: PropTypes.array.isRequired,
    lines: PropTypes.array,

    dataKey: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    trackHeight: PropTypes.number,
    label: PropTypes.bool,
    tooltips: PropTypes.bool,
    tooltipContent: PropTypes.func,
    brush: PropTypes.bool,
    brushRange: PropTypes.array,
    range: PropTypes.array,
    onBrush: PropTypes.func,
    onMouseover: PropTypes.func,
    onClick: PropTypes.func,
};

//
// const defaultProps = {
//     styles: {
//         label: {
//             fontFamily: 'Comic Sans MS',
//             color: 'green'
//         },
//         input: {
//             background: '#ddd',
//             border: '1px solid red'
//         }
//     }
// };

class Timeline extends React.Component {

    componentDidMount() {
        this.key = Math.round(Math.random() * 100000).toFixed(0) + '-' + new Date().getTime();
        this.createTimeline();
    }


    componentDidUpdate() {
    }

    componentWillReceiveProps(nextProps, nextState) {
        if (!equals(nextProps.data, this.props.data)
                || !equals(nextProps.lines, this.props.lines)) {
            this.timelineFn
                .data(nextProps.data)
                .lines(nextProps.lines)
                .update()
                .updateRange(nextProps.range)
                .updateBrushRange(nextProps.brushRange)
                .redraw();
        }
        if (!equals(nextProps.range, this.props.range)){
            this.timelineFn.updateRange(nextProps.range);
        }
        if (!equals(nextProps.brushRange, this.props.brushRange)){
            this.timelineFn.updateBrushRange(nextProps.brushRange);
        }
    }

    createTimeline = () => {

        const {data, dataKey, lines, height, width, trackHeight, label, tooltips, tooltipContent, brush, brushRange, onBrush, onMouseover, onClick, range} = this.props;
        const config = {
            dataKey,
            width,
            height,
            trackHeight,
            label,
            tooltips,
            tooltipContent,
            brush,
            brushRange,
            range,
            onBrush,
            onMouseover,
            onClick
        };
        this.timelineFn = timeline(this.div, config).create(data, lines);
        this.timelineFn.redraw();
    };

    render() {
        return <div ref={input => this.div = input}/>
    }
}

Timeline.propTypes = propTypes;
// Timeline.defaultProps = defaultProps;

export default Timeline;
