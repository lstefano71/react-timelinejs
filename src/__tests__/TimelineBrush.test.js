import React from 'react';
import Timeline from '../Timeline';
import {mockData, mockRange} from './Timeline.tester';

test('render with brush', () => {

    const data = mockData({size: 3});
    const range = mockRange(data);
    const timeline = mount(<Timeline data={data} brush={true} brushRange={range}/>);
    expect(timeline.html()).toMatchSnapshot();

});