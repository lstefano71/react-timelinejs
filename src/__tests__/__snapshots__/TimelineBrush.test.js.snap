// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with brush 1`] = `
"
<div class=\\"timeline\\">
  <svg class=\\"svg\\"
       id=\\"svg-1\\"
       width=\\"700\\"
       height=\\"400\\"
  >
    <g transform=\\"translate(0,0)\\">
      <clipPath id=\\"chart-area\\">
        <rect width=\\"700\\"
              height=\\"400\\"
        >
        </rect>
      </clipPath>
      <g class=\\"chart\\"
         clip-path=\\"url(#chart-area)\\"
      >
        <g id=\\"band0\\"
           transform=\\"translate(0,0)\\"
        >
          <rect class=\\"band\\"
                width=\\"700\\"
                height=\\"380\\"
          >
          </rect>
          <path d=\\"M0 0 L0 380\\"
                class=\\"mouseover-line\\"
                style=\\"display: none;\\"
          >
          </path>
          <rect height=\\"16\\"
                width=\\"75\\"
                y=\\"364\\"
                class=\\"mouseover-label-background\\"
                style=\\"display: none;\\"
          >
          </rect>
          <text x=\\"0\\"
                y=\\"376\\"
                class=\\"mouseover-label\\"
                style=\\"display: none;\\"
          >
          </text>
          <svg y=\\"4\\"
               height=\\"19\\"
               class=\\"part interval\\"
               x=\\"660.7030050328602\\"
               width=\\"39.29699496713977\\"
          >
            <rect width=\\"100%\\"
                  height=\\"100%\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"10%\\"
                  y=\\"0\\"
                  width=\\"20%\\"
                  height=\\"19\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"40%\\"
                  y=\\"0\\"
                  width=\\"9.999999999999998%\\"
                  height=\\"19\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"55.00000000000001%\\"
                  y=\\"0\\"
                  width=\\"19.999999999999996%\\"
                  height=\\"19\\"
            >
            </rect>
            <text class=\\"intervalLabel\\"
                  x=\\"1\\"
                  y=\\"10\\"
            >
              event 2
            </text>
          </svg>
          <svg y=\\"4\\"
               height=\\"19\\"
               class=\\"part interval\\"
               x=\\"45.06531589189063\\"
               width=\\"24.427861736330136\\"
          >
            <rect width=\\"100%\\"
                  height=\\"100%\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"10%\\"
                  y=\\"0\\"
                  width=\\"20%\\"
                  height=\\"19\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"40%\\"
                  y=\\"0\\"
                  width=\\"9.999999999999998%\\"
                  height=\\"19\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"55.00000000000001%\\"
                  y=\\"0\\"
                  width=\\"19.999999999999996%\\"
                  height=\\"19\\"
            >
            </rect>
            <text class=\\"intervalLabel\\"
                  x=\\"1\\"
                  y=\\"10\\"
            >
              event 0
            </text>
          </svg>
          <svg y=\\"4\\"
               height=\\"19\\"
               class=\\"part interval\\"
               x=\\"0\\"
               width=\\"18.586416538512058\\"
          >
            <rect width=\\"100%\\"
                  height=\\"100%\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"10%\\"
                  y=\\"0\\"
                  width=\\"20%\\"
                  height=\\"19\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"40%\\"
                  y=\\"0\\"
                  width=\\"9.999999999999998%\\"
                  height=\\"19\\"
            >
            </rect>
            <rect class=\\"step custom_1\\"
                  x=\\"55.00000000000001%\\"
                  y=\\"0\\"
                  width=\\"19.999999999999996%\\"
                  height=\\"19\\"
            >
            </rect>
            <text class=\\"intervalLabel\\"
                  x=\\"1\\"
                  y=\\"10\\"
            >
              event 1
            </text>
          </svg>
          <svg class=\\"x brush\\"
               fill=\\"none\\"
               pointer-events=\\"all\\"
          >
            <rect class=\\"overlay\\"
                  pointer-events=\\"all\\"
                  cursor=\\"crosshair\\"
                  x=\\"0\\"
                  y=\\"4\\"
                  width=\\"700\\"
                  height=\\"376\\"
            >
            </rect>
            <rect class=\\"selection\\"
                  cursor=\\"move\\"
                  fill=\\"#777\\"
                  fill-opacity=\\"0.3\\"
                  stroke=\\"#fff\\"
                  shape-rendering=\\"crispEdges\\"
                  style
                  y=\\"0\\"
                  height=\\"380\\"
                  x=\\"174.99999778733135\\"
                  width=\\"349.99999557466265\\"
            >
            </rect>
            <rect class=\\"handle handle--e\\"
                  cursor=\\"ew-resize\\"
                  style
                  y=\\"-3\\"
                  height=\\"386\\"
                  x=\\"521.999993361994\\"
                  width=\\"6\\"
            >
            </rect>
            <rect class=\\"handle handle--w\\"
                  cursor=\\"ew-resize\\"
                  style
                  y=\\"-3\\"
                  height=\\"386\\"
                  x=\\"171.99999778733135\\"
                  width=\\"6\\"
            >
            </rect>
          </svg>
        </g>
        <g class=\\"axis\\"
           transform=\\"translate(0,380)\\"
           fill=\\"none\\"
           font-size=\\"10\\"
           font-family=\\"sans-serif\\"
           text-anchor=\\"middle\\"
        >
          <path class=\\"domain\\"
                stroke=\\"#000\\"
                d=\\"M0.5,6V0.5H700.5V6\\"
          >
          </path>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(22.92722871079924,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              09 AM
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(118.51451376600411,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              12 PM
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(214.101798821209,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              03 PM
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(309.68908387641386,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              06 PM
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(405.2763689316188,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              09 PM
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(500.86365398682364,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              Jul 15
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(596.4509390420285,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              03 AM
            </text>
          </g>
          <g class=\\"tick\\"
             opacity=\\"1\\"
             transform=\\"translate(692.0382240972334,0)\\"
          >
            <line stroke=\\"#000\\"
                  y2=\\"6\\"
            >
            </line>
            <text fill=\\"#000\\"
                  y=\\"9\\"
                  dy=\\"0.71em\\"
            >
              06 AM
            </text>
          </g>
        </g>
      </g>
    </g>
  </svg>
  <div class=\\"tooltip\\"
       id=\\"tooltip-1\\"
       style=\\"visibility: hidden;\\"
  >
  </div>
</div>
"
`;
